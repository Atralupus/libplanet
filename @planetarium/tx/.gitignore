coverage/
dist/
node_modules/
planetarium-tx-*.tgz
