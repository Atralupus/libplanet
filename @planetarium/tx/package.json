{
  "name": "@planetarium/tx",
  "private": true,
  "description": "Creating Libplanet transactions from JavaScript/TypeScript",
  "type": "module",
  "main": "./dist/index.js",
  "exports": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist/**/*"
  ],
  "engines": {
    "node": ">=19.0.0"
  },
  "scripts": {
    "build": "yarn && nanobundle build",
    "prepack": "yarn && yarn build",
    "dev": "yarn && dotnet build ../../Libplanet.Tools && vitest",
    "test": "yarn && yarn run -T tsc -p tsconfig.json && dotnet build ../../Libplanet.Tools && vitest run",
    "coverage": "yarn && vitest run --coverage"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planetarium/libplanet.git",
    "directory": "@planetarium/tx"
  },
  "keywords": [
    "libplanet"
  ],
  "author": "Planetarium (https://planetarium.dev/)",
  "license": "LGPL-2.1-or-later",
  "bugs": {
    "url": "https://github.com/planetarium/libplanet/labels/js"
  },
  "homepage": "https://github.com/planetarium/libplanet/tree/main/@planetarium/tx",
  "devDependencies": {
    "@types/node": "^18.14.0",
    "@vitest/coverage-c8": "^0.22.1",
    "@vitest/ui": "^0.22.1",
    "execa": "^6.1.0",
    "fast-check": "^3.1.2",
    "nanobundle": "^1.5.0",
    "vite": "^4.1.3",
    "vitest": "^0.22.1"
  },
  "dependencies": {
    "bencodex": "^0.1.1"
  }
}
