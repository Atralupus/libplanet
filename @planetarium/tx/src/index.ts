export { type Address, encodeAddress } from "./address.js";
export {
  type Currency,
  encodeCurrency,
  type FungibleAssetValue,
  encodeFungibleAssetValue,
  getCurrencyHash,
  getMajorUnit,
  getMinorUnit,
  getSign,
} from "./assets.js";
export type {
  TxMetadata,
  UnsignedTx,
} from "./tx/index.js";
export {
  encodeTxMetadata,
  encodeUnsignedTx,
} from "./tx/index.js";
export {
  signTx,
  encodeSignedTx,
} from "./tx/signed.js";
