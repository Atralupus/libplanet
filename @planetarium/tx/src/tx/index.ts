export type { TxMetadata } from "./metadata.js";
export { encodeTxMetadata } from "./metadata.js";
export type {
  UnsignedTxWithSystemAction,
  UnsignedTxWithCustomActions,
} from "./unsigned.js";
export type { SignedTx } from "./signed.js";
export {
  encodeUnsignedTxWithSystemAction,
  encodeUnsignedTxWithCustomActions,
} from "./unsigned.js";
export { signTx, encodeSignedTx } from "./signed.js";
