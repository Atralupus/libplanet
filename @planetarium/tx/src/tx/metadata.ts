import {
  BencodexDictionary,
  Dictionary,
  Key,
  Value,
} from "@planetarium/bencodex";
import { Address, encodeAddress, encodeAddressSet } from "../address.js";
import { BlockHash, encodeBlockHash } from "../blockhash.js";
import { encodePublicKey, PublicKey } from "../key.js";

/**
 * Represents an unsigned transaction without actions.  Corresponds to
 * Libplanet's `TxMetadata`.
 */
export interface TxMetadata {
  nonce: bigint;
  publicKey: PublicKey;
  signer: Address; // TODO: This field can be derived from publicKey.
  timestamp: Date;
  updatedAddresses: Set<Address>;
  genesisHash: BlockHash | null;
}

const NONCE_KEY = new Uint8Array([0x6e]); // 'n'
const SIGNER_KEY = new Uint8Array([0x73]); // 's'
const GENESIS_HASH_KEY = new Uint8Array([0x67]); // 'g'
const UPDATED_ADDRESSES_KEY = new Uint8Array([0x75]); // 'u'
const PUBLIC_KEY_KEY = new Uint8Array([0x70]); // 'p'
const TIMESTAMP_KEY = new Uint8Array([0x74]); // 't'

export function encodeTxMetadata(metadata: TxMetadata): Dictionary {
  const updatedAddresses = encodeAddressSet(metadata.updatedAddresses);
  const timestamp = metadata.timestamp.toISOString().replace(/Z$/, "000Z");
  const pairs: [Key, Value][] = [
    [NONCE_KEY, metadata.nonce],
    [SIGNER_KEY, encodeAddress(metadata.signer)],
    [UPDATED_ADDRESSES_KEY, updatedAddresses],
    [PUBLIC_KEY_KEY, encodePublicKey(metadata.publicKey)],
    [TIMESTAMP_KEY, timestamp],
  ];
  if (metadata.genesisHash !== null) {
    pairs.push([GENESIS_HASH_KEY, encodeBlockHash(metadata.genesisHash)]);
  }
  return new BencodexDictionary(pairs);
}
