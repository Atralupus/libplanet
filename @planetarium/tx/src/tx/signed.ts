import { BencodexDictionary, Dictionary, encode } from "@planetarium/bencodex";
import { Account, Address, Signature } from "@planetarium/account";
import {
  type UnsignedTxWithSystemAction,
  type UnsignedTxWithCustomActions,
  encodeUnsignedTxWithSystemAction,
  encodeUnsignedTxWithCustomActions,
} from "./unsigned.js";
import { bytesEqual } from "../bytes.js";

const SIGNATURE_KEY = new Uint8Array([0x53]); // 'S'

export type SignedTx<
  T extends UnsignedTxWithSystemAction | UnsignedTxWithCustomActions,
> = T & { signature: Signature };

export async function signTx(
  tx: UnsignedTxWithSystemAction | UnsignedTxWithCustomActions,
  signAccount: Account,
): Promise<SignedTx<typeof tx>> {
  if (
    !bytesEqual(tx.publicKey, signAccount.publicKey.toBytes("uncompressed"))
  ) {
    throw new Error("Public keys in the tx and the signAccount are mismatched");
  } else if (
    !bytesEqual(tx.signer, Address.deriveFrom(signAccount.publicKey).toBytes())
  ) {
    throw new Error("The transaction signer does not match to the signAccount");
  }
  const payload =
    "systemAction" in tx
      ? encodeUnsignedTxWithSystemAction(tx)
      : encodeUnsignedTxWithCustomActions(tx);
  const signature = await signAccount.sign(encode(payload));
  return {
    ...tx,
    signature,
  };
}

export function encodeSignedTx<
  T extends UnsignedTxWithSystemAction | UnsignedTxWithCustomActions,
>(tx: SignedTx<T>): Dictionary {
  const dict =
    "systemAction" in tx
      ? encodeUnsignedTxWithSystemAction(tx)
      : encodeUnsignedTxWithCustomActions(tx);
  const sig = tx.signature.toBytes();
  return new BencodexDictionary([...dict, [SIGNATURE_KEY, sig]]);
}
