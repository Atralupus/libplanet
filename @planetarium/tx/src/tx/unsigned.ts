import { BencodexDictionary, Dictionary, Value } from "@planetarium/bencodex";
import { encodeTxMetadata, TxMetadata } from "./metadata.js";

const ACTION_KEY = new Uint8Array([0x61]); // 'a'

export interface UnsignedTx extends TxMetadata {
  actions: Value[];
}

/**
 * Encodes an unsigned transaction.
 * @param tx An unsigned transaction.
 * @returns An encoded transaction.
 */
export function encodeUnsignedTx(metadata: UnsignedTx): Dictionary {
  return new BencodexDictionary([
    ...encodeTxMetadata(metadata),
    [ACTION_KEY, metadata.actions],
  ]);
}
