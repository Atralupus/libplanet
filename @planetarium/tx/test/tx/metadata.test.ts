import { encode } from "bencodex";
import { expect, test } from "vitest";
import { encodeTxMetadata } from "../../src/tx/metadata";
import { key1 } from "./fixtures";

test("encodeTxMetadata", () => {
  const encoded = encodeTxMetadata({
    nonce: 123n,
    publicKey: key1.public,
    signer: key1.address,
    timestamp: new Date(2022, 4, 23, 10, 2, 0),
    updatedAddresses: new Set(),
    genesisHash: null,
  });
  expect(encoded).toMatchSnapshot();
  expect(encode(encoded)).toMatchSnapshot();
});
