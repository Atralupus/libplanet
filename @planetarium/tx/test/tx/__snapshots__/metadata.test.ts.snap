// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encodeTxMetadata 1`] = `
"BencodexDictionary(5) {
  Uint8Array(1) [
    110
  ] => 123n,
  Uint8Array(1) [
    115
  ] => Uint8Array(20) [
    38,
    131,
    68,
    186,
    70,
    230,
    202,
    42,
    138,
    80,
    150,
    86,
    85,
    72,
    185,
    1,
    139,
    198,
    135,
    206
  ],
  Uint8Array(1) [
    117
  ] => [],
  Uint8Array(1) [
    112
  ] => Uint8Array(33) [
    2,
    0,
    224,
    39,
    9,
    204,
    12,
    5,
    29,
    193,
    5,
    24,
    140,
    69,
    74,
    46,
    126,
    247,
    179,
    107,
    133,
    218,
    52,
    82,
    157,
    58,
    188,
    25,
    104,
    22,
    124,
    245,
    79
  ],
  Uint8Array(1) [
    116
  ] => '2022-05-23T01:02:00.000000Z'
}"
`;

exports[`encodeTxMetadata 2`] = `
Uint8Array [
  100,
  49,
  58,
  110,
  105,
  49,
  50,
  51,
  101,
  49,
  58,
  112,
  51,
  51,
  58,
  2,
  0,
  224,
  39,
  9,
  204,
  12,
  5,
  29,
  193,
  5,
  24,
  140,
  69,
  74,
  46,
  126,
  247,
  179,
  107,
  133,
  218,
  52,
  82,
  157,
  58,
  188,
  25,
  104,
  22,
  124,
  245,
  79,
  49,
  58,
  115,
  50,
  48,
  58,
  38,
  131,
  68,
  186,
  70,
  230,
  202,
  42,
  138,
  80,
  150,
  86,
  85,
  72,
  185,
  1,
  139,
  198,
  135,
  206,
  49,
  58,
  116,
  117,
  50,
  55,
  58,
  50,
  48,
  50,
  50,
  45,
  48,
  53,
  45,
  50,
  51,
  84,
  48,
  49,
  58,
  48,
  50,
  58,
  48,
  48,
  46,
  48,
  48,
  48,
  48,
  48,
  48,
  90,
  49,
  58,
  117,
  108,
  101,
  101,
]
`;
